VideoPlayer/RendererVAAPIGL*: initialize output parameters

MaxKellermann opened the pull request:
If vaInitialize() fails, then the "general" and "deepColor" parameters
are never initialized, which looks like this in valgrind:

```
 Conditional jump or move depends on uninitialised value(s)
    at 0x15C3221: CWinSystemX11GLContext::RefreshGLContext(bool) (WinSystemX11GLContext.cpp:289)
    by 0x15C2AA7: CWinSystemX11GLContext::SetWindow(int, int, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int*) (WinSystemX11GLContext.cpp:153)
    by 0x15B997B: CWinSystemX11::SetFullScreen(bool, RESOLUTION_INFO&, bool) (WinSystemX11.cpp:289)
    by 0x15C2D8E: CWinSystemX11GLContext::SetFullScreen(bool, RESOLUTION_INFO&, bool) (WinSystemX11GLContext.cpp:207)
    by 0x15B8EE1: CWinSystemX11::CreateNewWindow(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, RESOLUTION_INFO&) (WinSystemX11.cpp:105)
    by 0x15C2C24: CWinSystemX11GLContext::CreateNewWindow(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, RESOLUTION_INFO&) (WinSystemX11GLContext.cpp:175)
    by 0x1DCF226: CApplication::InitWindow(RESOLUTION) (Application.cpp:723)
    by 0x1DCEDF1: CApplication::CreateGUI() (Application.cpp:674)
    by 0x1A1B189: XBMC_Run (xbmc.cpp:65)
    by 0x13DE3D1: main (main.cpp:108)
  Uninitialised value was created by a stack allocation
    at 0x15C2EF2: CWinSystemX11GLContext::RefreshGLContext(bool) (WinSystemX11GLContext.cpp:252)
```

To fix this, we need to let the caller know whether the registration
was successful.

Rechi commented:
Please give this PR (and the commit) a meaningful title.

pkerling commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/RendererVAAPIGL.cpp:
```diff
@@ -40,13 +40,13 @@ CBaseRenderer* CRendererVAAPI::Create(CVideoBuffer *buffer)
+    return false;
   }
 
   general = deepColor = false;
```
Wouldn't moving this a few lines up also solve this?

MaxKellermann replied:
How?

Rechi replied:
Your stack trace tells that `general` isn't initialised. So move it up to also set them to false if `vaInitialize` fails.

MaxKellermann replied:
So by "this" you meant the assignment of the two variables? Yes, one could do that, but since those parameters were not documented, I could not figure out their meaning, and was not sure whether this assignment after a `vaInitialize()` failure was a valid state. Is it?

pkerling replied:
By "this" I meant the line I commented on. Sorry if it wasn't obvious.
The variables describe whether vaapi is available (`general`) and whether deep color rendering is available (`deepColor`). So yes, having both set to `false` is the expected state when vaapi initialization fails.

MaxKellermann replied:
Oh, then this code was more inconsistent and confusing than I thought. Because the line you commented on wasn't even necessary at all previously - `CVaapi?Texture::TestInterop()` already does that. That's what you get for putting return values in reference parameters...
The real fix (without fixing the API design) would be to move the assignments into the `vaInitialize()` error branch.

pkerling replied:
> The real fix (without fixing the API design) would be to move the assignments into the vaInitialize() error branch.

I think moving the assignment to the top of the function is clearer and doesn't hurt.

@FernetMenta I also very much dislike the non-const reference parameters. Can we change this?

MaxKellermann replied:
My idea for a proper API design: throw exception on error, and on success, return additional information (e.g. `deepColor`). But never return values in reference parameters.

Rechi commented on xbmc/windowing/X11/WinSystemX11GLContext.cpp:
```diff
@@ -285,8 +285,7 @@ bool CWinSystemX11GLContext::RefreshGLContext(bool force)
         X11::VaapiProxyConfig(m_vaapiProxy.get(), GetDisplay(),
                               static_cast<CGLContextEGL*>(m_pGLContext)->m_eglDisplay);
         bool general, deepColor;
```
Shouldn't the variables set to false here?
Otherwise the `#if !defined(HAVE_LIBVA)` version of `X11::VAAPIRegisterRender` has to set them to false too

MaxKellermann replied:
One could do that, but only `general` needs to be initialized, right? If `!general`, then `deepColor` should not ever be evaluated.

Rechi commented on xbmc/windowing/gbm/WinSystemGbmGLESContext.cpp:
```diff
@@ -60,9 +60,7 @@ bool CWinSystemGbmGLESContext::InitWindowSystem()
   bool general, deepColor;
```
same

MaxKellermann committed:
VideoPlayer/RendererVAAPIGL*: initialize output parameters

If vaInitialize() fails, then the "general" and "deepColor" parameters
are never initialized, which looks like this in valgrind:

```
 Conditional jump or move depends on uninitialised value(s)
    at 0x15C3221: CWinSystemX11GLContext::RefreshGLContext(bool) (WinSystemX11GLContext.cpp:289)
    by 0x15C2AA7: CWinSystemX11GLContext::SetWindow(int, int, bool, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int*) (WinSystemX11GLContext.cpp:153)
    by 0x15B997B: CWinSystemX11::SetFullScreen(bool, RESOLUTION_INFO&, bool) (WinSystemX11.cpp:289)
    by 0x15C2D8E: CWinSystemX11GLContext::SetFullScreen(bool, RESOLUTION_INFO&, bool) (WinSystemX11GLContext.cpp:207)
    by 0x15B8EE1: CWinSystemX11::CreateNewWindow(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, RESOLUTION_INFO&) (WinSystemX11.cpp:105)
    by 0x15C2C24: CWinSystemX11GLContext::CreateNewWindow(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, RESOLUTION_INFO&) (WinSystemX11GLContext.cpp:175)
    by 0x1DCF226: CApplication::InitWindow(RESOLUTION) (Application.cpp:723)
    by 0x1DCEDF1: CApplication::CreateGUI() (Application.cpp:674)
    by 0x1A1B189: XBMC_Run (xbmc.cpp:65)
    by 0x13DE3D1: main (main.cpp:108)
  Uninitialised value was created by a stack allocation
    at 0x15C2EF2: CWinSystemX11GLContext::RefreshGLContext(bool) (WinSystemX11GLContext.cpp:252)
```

To fix this, move the output parameter initialization to the top of
the Register() method.

MaxKellermann commented:
I've pushed a new commit implementing @pkerling's idea.

MaxKellermann commented:
@Rechi that additional bug you found may be correct, but this shouldn't delay merging this PR, because your problem is mostly orthogonal to this one.

Rechi commented:
If you merge my PR into your branch, this PR can be merged after jenkins has finished (~ 30 mins). I think we can deal with that delay.

FernetMenta commented:
thanks!
